1-Draft/RFC-Support for .NET extension methods.md

```diff
@@ -0,0 +1,82 @@
+
+This RFC proposes changes to language syntax and runtime binding to enable the calling of generic .NET methods, and finding appropriate extension methods based on the type of the extension methods first parameter.
+
+## Motivation
+
+Extension methods enable you to "add" methods to existing types without creating a new derived type, recompiling, or otherwise modifying the original type. Extension methods are a special kind of static method, but they are called as if they were instance methods on the extended type. For client code written in PowerShell, there would be no apparent difference between calling an extension method and the methods that are actually defined in a type.
+
+There are lots of useful extension methods in .NET already written, many of which are in the area of processing sets of objects and where the expressiveness/performance ratio, from a PowerShell perspective, is quite favorable.
+
+### Goals
+
+Simplicity of use
+-	Should be as easy as possible to call from PowerShell.
+    -	Type parameters will be deduced to the most specific type possible
+```powershell
+    # System.Linq.Enumerable.Select[TInput, TResult] takes two generic type parameters
+    using namespace System.Linq    
+    $a.Select({param([int] $i) $i.ToString()}) # [int,object] will be deduced.
+
+    [func[int,string]] $f = {param([int]$i) ($i * 2).ToString()}
+    $a.Select($f) # [int,string] will be deduced.
+    
+    $a.Select(param($i) $i * 2) #  [object,object] will be deduced.
+```
+
+Simplication of 'Magic' methods
+The `Foreach` and `Where` methods that are added to collections are a hack which could be solved more elegantly by making extension methods of the EnumerableOps methods.
+By doing this, they wouldn't need to be special cased. Normal method resolution would make it work, and would also make `List[T].Foreach` not hide the extension methods.
+
+```csharp
+public static class EnumerableOps {
+    static IEnumerable<T> Foreach<T>(this IEnumerable<T> c, ScriptBlock sb){}
+    static IEnumerable<T> Foreach<T>(this IEnumerable<T> c, ScriptBlock sb, WhereOperatorSelectionMode mode){}
+    static IEnumerable<T> Foreach<T>(this IEnumerable<T> c, ScriptBlock sb, WhereOperatorSelectionMode mode, int count){}
+    static IEnumerable<T> Foreach<T>(this IEnumerable<T> c, object expression, object[] args){}
+
+    static IEnumerable<T> Where<T>(this IEnumerable<T> c, ScriptBlock sb){}
+    static IEnumerable<T> Where<T>(this IEnumerable<T> c, ScriptBlock sb, WhereOperatorSelectionMode mode){}
+    static IEnumerable<T> Where<T>(this IEnumerable<T> c, ScriptBlock sb, WhereOperatorSelectionMode mode, int count){}
+}
+
+```
+
+
+## Specification
+
+Extension methods are choosen from all methods on all loaded types where the methods has the ExtensionAttribute.
+From these, the resulting set is limited to the types whose namespace are in use in the context where the call is made.
+From these, a best fit is selected based on parameters.
```

Comment:
Are you proposing that `using namespace` directives should always be specified with an extension method invocation?

I think it's more intuitive to not require `using namespace`. However, when namespaces are specified, extension methods from them are more preferred during resolution. 
For example, `obj.GetAcl(...)` would make powershell search from all assemblies (loaded and loadable TPAs) for a match of extension method named `GetAcl`, and if we find equally good matching from `System.IO.FileSystemAccessControl.dll` and `MyAssembly.dll`, powershell reports error about the ambiguity.

But if the `using` directive is specified as below, then the extension method from under `System.IO` should be chosen.
```
using System.IO
obj.GetAcl(...)
```

Reply:
Now looking back, I'm also inclined to make `using namespace` a requirement for extension method resolution to kick in. I think this will make the script more portable. If powershell is asked to resolve an extension method without being given any hints, then it will just go through all loaded assemblies and discoverable TPAs, which will result in two problems:
1. We don't control what assemblies are searched in what order, so the same extension method may be resolved differently in a different session.
2. If we stop at the first found compatible extension method, then it's possible the resolve method is not what the user wants.

Reply:
Given a namespace, powershell still needs to look for all loaded and discoverable assemblies to find all types from the namespace, because the same namespace might be used in multiple assemblies (`Microsoft.PowerShell.Commands` is an example, which is used in `S.M.A.dll`, `M.P.Command.Utility.dll`, `M.P.Command.Management.dll` and etc.).

C# is able to discover the most suitable overloads from the given namespace that are from two assemblies.